tests/cases/sftp_test.py

```diff
@@ -0,0 +1,201 @@
+        self.assertEqual(set(sftpClient.listdir('/user/admin')), {'Public', 'Private'})
+        self.assertEqual(
+            set(sftpClient.listdir('/collection')), {'public collection', 'private collection'})
+
+        self.assertEqual(sftpClient.listdir('/user/regularuser/Private'), ['test.txt'])
+        self.assertEqual(sftpClient.listdir('/user/regularuser/Private/test.txt'), ['test.txt'])
+
+        with six.assertRaisesRegex(self, IOError, 'No such file'):
+            sftpClient.listdir('/user/nonexistent')
+
+        with six.assertRaisesRegex(self, IOError, 'No such file'):
+            sftpClient.file('/user/regularuser/Private')
+
+        # Read a file using small enough buf size to require multiple chunks internally.
+        file = sftpClient.file('/user/regularuser/Private/test.txt/test.txt', 'r', bufsize=4)
+        self.assertEqual(file.read(2), b'he')
+        self.assertEqual(file.read(), b'llo world')
+
+        # Test stat capability
+        info = sftpClient.stat('/user/regularuser/Private')
+        self.assertTrue(stat.S_ISDIR(info.st_mode))
+        self.assertFalse(stat.S_ISREG(info.st_mode))
+        self.assertEqual(info.st_mode & 0o777, 0o777)
+
+        info = sftpClient.stat('/user/regularuser/Private/test.txt/test.txt')
+        self.assertFalse(stat.S_ISDIR(info.st_mode))
+        self.assertTrue(stat.S_ISREG(info.st_mode))
+        self.assertEqual(info.st_size, 11)
+        self.assertEqual(info.st_mode & 0o777, 0o777)
+
+        # File stat implementations should agree
+        info = file.stat()
+        self.assertFalse(stat.S_ISDIR(info.st_mode))
+        self.assertTrue(stat.S_ISREG(info.st_mode))
+        self.assertEqual(info.st_size, 11)
+        self.assertEqual(info.st_mode & 0o777, 0o777)
+
+        sftpClient.close()
+        client.close()
+
+        # Test anonymous access
+        client = paramiko.SSHClient()
+        client.load_system_host_keys()
+        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+        client.connect(
+            'localhost', TEST_PORT, username='anonymous', password='', look_for_keys=False)
+        sftpClient = client.open_sftp()
+
+        # Only public data should be visible
+        self.assertEqual(set(sftpClient.listdir('/user')), {'admin', 'regularuser'})
+        self.assertEqual(sftpClient.listdir('/collection'), ['public collection'])
+        self.assertEqual(sftpClient.listdir('/user/admin'), ['Public'])
+
+        # Make sure the client cannot distinguish between a resource that does not exist
+        # vs. one they simply don't have read access to.
+        with six.assertRaisesRegex(self, IOError, 'No such file'):
+            sftpClient.listdir('/user/regularuser/Private')
+
+        with six.assertRaisesRegex(self, IOError, 'No such file'):
+            sftpClient.file('/user/regularuser/Private/test.txt/test.txt', 'r')
+
+        sftpClient.close()
+        client.close()
```

Comment:
The only reason I didn't was because they both use the same test dataset and I didn't want to have to recreate it in the database twice. If you think it's worthwhile, I can break it up.